packages/integrations/vercel/src/serverless/entrypoint.ts

```diff
@@ -1,35 +1,21 @@
 import type { SSRManifest } from 'astro';
-import { App } from 'astro/app';
-import { applyPolyfills } from 'astro/app/node';
+import { applyPolyfills, NodeApp } from 'astro/app/node';
 import type { IncomingMessage, ServerResponse } from 'node:http';
-
 import { ASTRO_LOCALS_HEADER } from './adapter.js';
-import { getRequest, setResponse } from './request-transform.js';
 
 applyPolyfills();
 
 export const createExports = (manifest: SSRManifest) => {
-	const app = new App(manifest);
-
+	const app = new NodeApp(manifest);
 	const handler = async (req: IncomingMessage, res: ServerResponse) => {
-		let request: Request;
-
-		try {
-			request = await getRequest(`https://${req.headers.host}`, req);
-		} catch (err: any) {
-			res.statusCode = err.status || 400;
-			return res.end(err.reason || 'Invalid request body');
-		}
-
-		let routeData = app.match(request);
-		let locals = {};
-		if (request.headers.has(ASTRO_LOCALS_HEADER)) {
-			let localsAsString = request.headers.get(ASTRO_LOCALS_HEADER);
-			if (localsAsString) {
-				locals = JSON.parse(localsAsString);
-			}
-		}
-		await setResponse(app, res, await app.render(request, { routeData, locals }));
+		const clientAddress = req.headers['x-forwarded-for'] as string | undefined;
+		const localsHeader = req.headers[ASTRO_LOCALS_HEADER]
+		const locals =
+			typeof localsHeader === "string" ? JSON.parse(localsHeader)
+			: Array.isArray(localsHeader) ? JSON.parse(localsHeader[0])
```

Comment:
Can you add a comment that explains why the locals header is an array?

Reply:
That's just its type. Node.js headers yield an array when the same header appears multiple times.

Reply:
Then why do we check if it's a string using `typeof`? Sorry, I might not know what's going on

Reply:
It yields a string when the header only appears a single time.